persistent/Database/Persist/TH/Internal.hs

```diff
@@ -280,9 +280,13 @@ embedEntityDefsMap existingEnts rawEnts =
 -- In 2.13.0.0, this was changed to splice in @['UnboundEntityDef']@
 -- instead of @['EntityDef']@.
 --
--- @since 2.5.3
+-- @since 2.16.0.0
 parseReferences :: PersistSettings -> [(Maybe SourceLoc, Text)] -> Q Exp
-parseReferences ps s = lift $ parse ps s
+parseReferences ps s = do
+  let cpr = parse ps s
+  case renderErrors cpr of
+    Nothing -> lift $ cumulativeData cpr
+    Just errorMessage -> fail errorMessage
```

Comment:
If you make `renderErrors` operate on `[EntityError]`, then this line should be:

```suggestion
  case cpr of
    Right res -> lift res
    Left errors -> fail $ renderErrors errors
```

One article I recommend is [Parse, Don't Validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) - which goes into why it's good to use datatypes like `Either` etc to refine information we have about types. You can think of `A -> Maybe B` as separating `A` into `B` and `Nothing`, which removes all hope of understanding what made the `A` a not-`B`. Instead, doing `A -> Either C B` lets you retain that information, and then use case-splitting to recover that information